src/engine/physics/FindGround.cpp

```diff
@@ -0,0 +1,29 @@
+#include "physics/FindGround.hpp"
+
+namespace lei3d
+{
+    FindGround::FindGround(btRigidBody* body) : m_Me(body) {}
+
+    btScalar FindGround::addSingleResult(btManifoldPoint& cp,
+                                    const btCollisionObjectWrapper* colObj0, int partId0, int index0,
+                                    const btCollisionObjectWrapper* colObj1, int partId1, int index1)
+    {
+        // We can use this to also check what m_collisionObject2 is to trigger things / change how we react to colliding with it
+        if (colObj0->m_collisionObject == m_Me && !m_HaveGround) {
+            const btTransform& transform = m_Me->getWorldTransform();
+            // Orthonormal basis (just rotations) -> can just transform to invert
+            btMatrix3x3 invBasis = transform.getBasis();
+            btVector3 localPoint = invBasis * (cp.m_positionWorldOnB - transform.getOrigin());
+            localPoint[1] += m_ShapeHalfHeight;
+            float r = localPoint.length();
```

Comment:
Very minor optimization, but you can use length2 to get the square of the radius and then compare it to m_ShapeRadius * m_ShapeRadius and m_RadiusThreshold * m_RadiusThreshold, esp. since I'm assuming ground checks are ran every frame. Also another point, but it feels weird to call everything radius when we're not necessarily talking about spheres.

Reply:
Right now this ground check is seemingly broken anyways so I will implement this after I fix the ground check